mllib/src/main/scala/org/apache/spark/ml/feature/UnivariateFeatureSelector.scala

```diff
@@ -0,0 +1,421 @@
+ *    positive rate of selection.
+ *  - `fdr` uses the <a href=
+ *  "https://en.wikipedia.org/wiki/False_discovery_rate#Benjamini.E2.80.93Hochberg_procedure">
+ *  Benjamini-Hochberg procedure</a>
+ *    to choose all features whose false discovery rate is below a threshold.
+ *  - `fwe` chooses all features whose p-values are below a threshold. The threshold is scaled by
+ *    1/numFeatures, thus controlling the family-wise error rate of selection.
+ *
+ * By default, the selection mode is `numTopFeatures`, with the default selectionThreshold
+ * sets to 50.
+ */
+@Since("3.1.1")
+final class UnivariateFeatureSelector @Since("3.1.1")(@Since("3.1.1") override val uid: String)
+  extends Estimator[UnivariateFeatureSelectorModel] with UnivariateFeatureSelectorParams
+    with DefaultParamsWritable {
+
+  @Since("3.1.1")
+  def this() = this(Identifiable.randomUID("UnivariateFeatureSelector"))
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setSelectionMode(value: String): this.type = set(selectionMode, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setSelectionThreshold(value: Double): this.type = set(selectionThreshold, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setFeatureType(value: String): this.type = set(featureType, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setLabelType(value: String): this.type = set(labelType, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /** @group setParam */
+  @Since("3.1.1")
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  @Since("3.1.1")
+  override def fit(dataset: Dataset[_]): UnivariateFeatureSelectorModel = {
+    transformSchema(dataset.schema, logging = true)
+    val spark = dataset.sparkSession
+    import spark.implicits._
+
+    val numFeatures = MetadataUtils.getNumFeatures(dataset, $(featuresCol))
+
+    require(isSet(featureType) && isSet(labelType), "featureType and labelType need to be set")
+    val resultDF = ($(featureType), $(labelType)) match {
+      case ("categorical", "categorical") =>
+        ChiSquareTest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
+      case ("continuous", "categorical") =>
+        ANOVATest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
+      case ("continuous", "continuous") =>
+        FValueTest.test(dataset.toDF, getFeaturesCol, getLabelCol, true)
+      case _ =>
+        throw new IllegalArgumentException(s"Unsupported combination:" +
+          s" featureType=${$(featureType)}, labelType=${$(labelType)}")
+    }
+
+    def getTopIndices(k: Int): Array[Int] = {
+      resultDF.sort("pValue", "featureIndex")
+        .select("featureIndex")
+        .limit(k)
+        .as[Int]
+        .collect()
+    }
+
+    val indices = $(selectionMode) match {
```

Comment:
Let's define a function like:
```
def selectIndicesFromPValues(selectionMode: String, selectionThreshold: Double): Array[Int]
```
and then add unit test on this.
especially for the fdr case

Reply:
@WeichenXu123 
I guess you want me to have more tests to verify fdr?  Is It OK with you that I add more tests for fdr and keep the methods as is? The other four selection methods are very straightforward and I guess we don't need more tests.